### PR TITLE
[Dev] fix: scale DSA indexer scoring backward for sparse indexer loss in FLOPs

### PR DESCRIPTION
**Pull request by niyunsheng**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Follow-up to #6753. That PR charged the DSA lightning-indexer scoring pass
`fwd + dq + dk = 3x` whenever `dsa_indexer_loss_coeff > 0`, regardless of
which KL variant is used. With `--dsa-indexer-use-sparse-loss` the KL is
taken over the selected top-k keys only, so the score gradient is zero
outside them and dq/dk of the scoring GEMM only span `sum_i(min(i, topk))`
pairs instead of the dense causal `L^2 / 2`. The fused cuDNN backend
(`_compute_sparse_indexer_loss_and_grads` -> `_run_sparse_indexer_backward`)
executes exactly this sparse backward. #6753 acknowledged this in the
docstring ("which the 3x does not model") but did not account for it.

Since the indexer's dense O(L^2) scoring dominates the quadratic term at
long context (the main attention core is already top-k scaled), the
overcount is close to 3x on that term. GLM-5.2 example (indexer 32 heads x
128, topk 2048, 128K sequence): indexer scoring coefficient drops from
`3 * 2048` to `(1 + 2 * 0.031) * 2048` per layer.

## Changes

- `_dsa_indexer_flops` takes `dsa_indexer_use_sparse_loss` and
  `sparse_core_scale` (the same top-k / dense pair ratio
  `_dsa_sparse_core_scale` applies to core attention). Scoring expansion:
  - indexer loss off: 1x (unchanged)
  - dense KL (`dsa_indexer_use_sparse_loss=False`, default): 3x (unchanged)
  - sparse KL (`dsa_indexer_use_sparse_loss=True`): `1 + 2 * sparse_core_scale`
- Parameters renamed to match the corresponding `args`
  (`dsa_indexer_loss_coeff`, `dsa_indexer_use_sparse_loss`).
- Caller computes `_dsa_sparse_core_scale` once and shares it between core
  attention and the indexer.
- Docstring rewritten to describe both loss variants; the reference PyTorch
  path multiplies a dense zero-padded gradient, which is an implementation
  detail and, like dense-masked reference attention, not part of the model
  FLOPs count.

Default configs (`dsa_indexer_use_sparse_loss=False`) produce identical
numbers to before.

## Tests

`tests/unit_tests/test_num_floating_point_operations.py`:
- golden calculator follows the same rule
- `test_sparse_indexer_loss_shrinks_scoring_backward[256|8192]`: sparse ==
  dense at `L <= topk`; strictly smaller past topk and still above a frozen
  indexer
- `test_sparse_loss_ignored_without_indexer_loss`: flag is a no-op when the
  indexer loss is off
- `test_indexer_flops_sparse_loss_expansion`: direct check of the 1x / 3x /
  `1 + 2 * scale` factors

Ran the file in the CI `dev` container on a GB200 node:
`57 passed, 8 skipped` (skips are the 8-rank `TestAccumulatorTopology`
cases on a 4-GPU node).

## Not covered

The teacher-side attention score recompute inside the indexer KL loss
(dense over all keys, or top-k only under the sparse variant) is still not
counted, consistent with #6753's rule of excluding auxiliary-loss work.

### CODE REVIEW

File: megatron/training/training.py [hunk 1]
```diff
@@ -423,7 +421,15 @@ def _dsa_sparse_core_scale(total_real_tokens, seqlen_squared_sum, dsa_indexer_to
...
 
 def _dsa_indexer_flops(
-    *, hidden_size, q_lora_rank, n_heads, head_dim, num_indexer_layers, indexer_loss_coeff
+    *,
+    hidden_size,
+    q_lora_rank,
+    n_heads,
+    head_dim,
+    num_indexer_layers,
+    dsa_indexer_loss_coeff,
```
Comment (buptzyb) [inline megatron/training/training.py]: I noticed that we actually don't care about this value. We just care it's 0 or >0. Can we replace this with a boolean argument like `dsa_indexer_loss_enabled` or `dsa_indexer_frozen`?

btw you added a `dsa_` prefix to this argument. What the reason for this change?

Comment (niyunsheng) [reply megatron/training/training.py]: Updated in 7eace39b3. The helper now accepts `dsa_indexer_loss_enabled: bool`, derived from `(args.dsa_indexer_loss_coeff or 0.0) > 0` at the call site. The training argument remains a float because its magnitude scales the KL-loss gradients. Updated the helper tests accordingly.

I added the `dsa_` prefix to align with the corresponding args names; it wasn't a functional change.

**Review by buptzyb — APPROVED** [review]

**Comment by buptzyb** [comment]
Can we make this and #6753 share the same main counterpart? If it's not merged yet.

**Comment by niyunsheng** [comment]
Yes, both #6753 and this PR are already included in the same main counterpart: https://github.com/NVIDIA/Megatron-LM/pull/7043. It is still open.

**Review by yaox12 — APPROVED** [review]
